server/src/internalClusterTest/java/org/opensearch/cluster/coordination/NodeJoinLeftIT.java

```diff
@@ -0,0 +1,374 @@
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        });
+        // Toggle to succeed/fail the followerchecker to simulate the initial node leaving.
+        AtomicBoolean succeedFollowerChecker = new AtomicBoolean();
+
+        // Simulate followerchecker failure on 1 node when succeedFollowerChecker is false
+        FollowerCheckerBehaviour simulatedFailureBehaviour = new FollowerCheckerBehaviour(() -> {
+            if (succeedFollowerChecker.get()) {
+                return;
+            }
+            try {
+                Thread.sleep(10);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            throw new NodeHealthCheckFailureException("fake followerchecker failure simulated by test to repro race condition");
+        });
+        MockTransportService redTransportService = (MockTransportService) internalCluster().getInstance(
+            TransportService.class,
+            redNodeName
+        );
+        redTransportService.addRequestHandlingBehavior(FOLLOWER_CHECK_ACTION_NAME, simulatedFailureBehaviour);
+
+        // Loop runs 10 times to ensure race condition gets reproduced
+        for (int i = 0; i < 10; i++) {
+            // Fail followerchecker by force to trigger node disconnect and node left
+            logger.info("--> simulating followerchecker failure to trigger node-left");
+            succeedFollowerChecker.set(false);
+            ClusterHealthResponse response1 = client().admin().cluster().prepareHealth().setWaitForNodes("2").get();
+            assertThat(response1.isTimedOut(), is(false));
+
+            // once we know a node has left, we can re-enable followerchecker to work normally and validate node rejoins
+            logger.info("--> re-enabling normal followerchecker and validating cluster is stable");
+            succeedFollowerChecker.set(true);
+            ClusterHealthResponse response2 = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+            assertThat(response2.isTimedOut(), is(false));
+
+            Thread.sleep(1000);
+            // checking again to validate stability and ensure node did not leave
+            ClusterHealthResponse response3 = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+            assertThat(response3.isTimedOut(), is(false));
+        }
+
+        succeedFollowerChecker.set(true);
+        ClusterHealthResponse response = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+        assertThat(response.isTimedOut(), is(false));
+
+        // assert that the right exception message showed up in logs
+        assertTrue(
+            "Expected IllegalStateException was not logged",
+            testAppender.containsExceptionMessage("IllegalStateException[cannot make a new connection as disconnect to node")
```

Comment:
Can we also verify the message logged on the datanode ?

Reply:
yes, will add this

Reply:
added this